### DISCUSSION TITLE
Question: how can I reduce bundle size?

### DISCUSSION THREAD

**Discussion by sandinmyjoints** [InboxSDK/InboxSDK#907] [Q&A] [answered]
After building for prod, I'm seeing InboxSDK is around 1MB of minified JS. I'm using only a few features related to composeViews, so I suspect a lot of that code is not needed. How can I exclude it from the prod bundle? I tried basic webpack tree-shaking with usedExports, but that did not make a difference.

**Comment by sandinmyjoints** [comment]
I tried this:

```diff
diff --git a/gulpfile.babel.ts b/gulpfile.babel.ts
index 0786e647..0dc0102d 100644
--- a/gulpfile.babel.ts
+++ b/gulpfile.babel.ts
@@ -153,6 +153,9 @@ async function webpackTask({
   const VERSION = await getVersion();
 
   const bundler = webpack({
+    experiments: {
+      outputModule: true,
+    },
     devtool:
       options.devtool === 'remote' ? false : options.devtool ?? 'source-map',
     entry: [
@@ -381,8 +384,8 @@ if (args.remote) {
     return webpackTask({
       entry: './src/inboxsdk-js/inboxsdk-NONREMOTE',
       destName: sdkFilename,
-      standalone: 'InboxSDK',
       afterBuild: setupExamples,
+      outputLibraryType: OutputLibraryType.ESM,
     });
   });
   gulp.task('remote', () => {
```

Then did `yarn gulp sdk`. As far as I can tell, the output (packages/core/inboxsdk.js) still does not look like an ESM module. 

I also noticed that packages/core/package.json has `sideEffects: true`, so I suspect that even if I got an ESM module built, nothing could be tree-shaken from it.

**Comment by wegry** [comment, marked as answer]
@Macil and I have talked a bit about making the switch from UMD to ESM for `@inboxsdk/core`. We would be open to a PR that makes the switch. There are a few overlapping issues that make that switch trickier.

## Tree Shaking and ESM

A few things potentially preventing us from tree-shaking the SDK better at the moment:

1. ~~We still use CJS as an intermediate format between Babel and Webpack. This might be the source of what you're seeing with the diff above. We _should_ use `modules: false` as a `@babel/preset-env` option, but our usage of `@babel/register` for our gulpfile interferes.~~ _edit: it looks like `modules: false` should function the same as the default `modules: 'auto'` because we use `babel-loader`_.
1. We use global namespaces for legacy remote-loaded versions of the SDK, e.g. `sdk.Widgets`. Even with ESM output, these assignments to a top level default export probably prevent tree shaking.
3. An alternative might be to just use modules files that you use. We do ship vanilla ESM versions of our modules. This was enabled as part of #848. Unfortunately, this mostly isn't workable at the moment as we have both CSS loading concerns that require webpack's runtime and Node global substitution (`process.env.*`) here and there.

## Code size

As far as code size goes, we've done a few things recently to try and knock down our output size. See https://bundlephobia.com/package/@inboxsdk/core@1.1.5 for a sampling of size changes over time.

- Delete dead code -> #887
- Update our browser targets -> #836 and #906
- Minify inlined CSS -> #896

I suspect there might be more dead code floating around. Deleting dead code will be much easier if we finish moving from Flow to TS. React is 13% of our SDK output. Swapping it out with something like preact would be a significant size savings.

**Comment by sandinmyjoints** [comment]
Thanks for the detailed reply! I have nothing helpful to offer regarding those blockers right now, except for re: point number 2, the overarching thought that separating legacy from current might be a useful direction to work towards (which is probably already on your radar).

Glad to know that reducing the size is already being tackled! It's a great project and I feel bad raising an issue without a solution to it. If I am able to learn more about these blockers and do anything useful, I will certainly follow up.

**Comment by wegry** [comment]
More potential ideas for things we could cut: 

1. Removing `GmailDriver.prototype.isUsingSyncApi` (see #993) and related checks.
2. Classes not part of our public API (Gmail*View) with non-public methods being moved to private ones would minify a good deal more.
